File: 1046-last-stone-weight/1046-last-stone-weight.cpp
struct compare {
    bool operator()(int a, int b) {
        return a < b;
    }
};

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int, vector<int>, compare> pq;
        
        for (int x: stones) {
            pq.push(x);
        }
        
        while (pq.size() > 1) {
            int x = pq.top();
            pq.pop();
            int y = pq.top();
            pq.pop();
            
            if (x != y) {
                pq.push(abs(y - x));
            }
        }
        
        return pq.empty() ? 0 : pq.top();
    }
};

File: 1046-last-stone-weight/1046-last-stone-weight.py

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        
        heap_ = [-s for s in stones]
        heapq.heapify(heap_)
        
        while len(heap_) > 1:
            x = heapq.heappop(heap_)
            y = heapq.heappop(heap_)
            
            if (x != y):
                heapq.heappush(heap_, x - y)
            
        return -heap_[0] if len(heap_) else 0 
